eclipse-collections/src/main/java/org/eclipse/collections/impl/set/strategy/mutable/UnifiedSetWithHashingStrategy.java

```diff
@@ -2132,6 +2160,73 @@ private T chainedGet(T key, ChainedBucket bucket)
         while (true);
     }
 
+    private T chainedAddOrReplace(T key, ChainedBucket bucket, int index)
+    {
+        do
+        {
+            if (this.nonNullTableObjectEquals(bucket.zero, key))
+            {
+                T previous = this.nonSentinel(bucket.zero);
+                bucket.zero = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                return previous;
+            }
+            if (bucket.one == null)
+            {
+                bucket.one = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                if (++this.occupied > this.maxSize)
+                {
+                    this.rehash();
+                }
+                return key;
+            }
+            if (this.nonNullTableObjectEquals(bucket.one, key))
+            {
+                T previous = this.nonSentinel(bucket.one);
+                bucket.one = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                return previous;
+            }
+            if (bucket.two == null)
+            {
+                bucket.two = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                if (++this.occupied > this.maxSize)
+                {
+                    this.rehash();
+                }
+                return key;
+            }
+            if (this.nonNullTableObjectEquals(bucket.two, key))
+            {
+                T previous = this.nonSentinel(bucket.two);
+                bucket.two = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                return previous;
+            }
+            if (bucket.three instanceof ChainedBucket)
+            {
+                bucket = (ChainedBucket) bucket.three;
+                continue;
+            }
+            if (bucket.three == null)
+            {
+                bucket.three = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                if (++this.occupied > this.maxSize)
+                {
+                    this.rehash();
+                }
+                return key;
+            }
+            if (this.nonNullTableObjectEquals(bucket.three, key))
+            {
+                T previous = this.nonSentinel(bucket.three);
+                bucket.three = UnifiedSetWithHashingStrategy.toSentinelIfNull(key);
+                return previous;
+            }
+
+            this.chainedAdd(key, index);
```

Comment:
This is a bit curious. Can we extract a method for adding a new value into a new chain?

Reply:
You mean instead of calling the chainedAdd()? I can extract it out.